test-app/backend/test/models/shift/shift-api.test.ts

```diff
@@ -10,118 +13,149 @@ const myShift = {
-    .send(myShift)
-    .expect(200)
-    .then((response) => {
-      expect(response.text).toBe('Shift has been updated');
+    test('Shift fails because max volunteers is 0', async () => {
+      await runValidationTestErrorNewShift({
+        ...BASE_SHIFT,
+        maxVolunteers: 0,
+      }, [
+        {
+          path: ['maxVolunteers'],
+          message: 'Number must be greater than or equal to 1',
+        },
+      ]);
     });
-});
+  });
 
-test('PUT  /api/shift/12; Shift fails to update because there is no title', async () => {
-  await request(app)
-    .put('/api/shift/12')
-    .send(noTitleShift)
-    .expect(400)
-    .then((response) => {
-      expect(response.text).toBe('Must include title');
+  describe('PUT /api/shift/:id', () => {
+    function getBaseSetup(e: string) {
+      return ({ shift }: IExpectedInput) => {
+        return {
+          req: request(app).put(e).send(shift),
+        };
+      };
+    }
+
+    test('Succcessful shift updated', async () => {
+      await runTest(BASE_SHIFT, getBaseSetup('/api/shift/12'), 200, ({ text }) => {
+        expect(text).toBe('Shift has been updated');
+      });
     });
-});
 
-test('PUT  /api/shift/13; Shift fails because start date is after end date', async () => {
-  await request(app)
-    .put('/api/shift/13')
-    .send(badDateShift)
-    .expect(400)
-    .then((response) => {
-      expect(response.text).toBe('Start/End dates are required and Start date must come before end date');
+    test('Shift fails because start date is after end date', async () => {
+      await runValidationTestError({
+        ...BASE_SHIFT,
+        start: BASE_SHIFT.end + 2,
+      }, getBaseSetup('/api/shift/12'), [
+        { path: [], message: 'start date must come before end date' },
+      ]);
     });
-});
 
-test('PUT  /api/shift/39; Shift fails because max volunteers is 0', async () => {
-  await request(app)
-    .put('/api/shift/39')
-    .send(badVolShift)
-    .expect(400)
-    .then((response) => {
-      expect(response.text).toBe('Max Volunteers is required and must be greater than 0');
+    test('Shift fails because max volunteers is 0', async () => {
+      await runValidationTestError({
+        ...BASE_SHIFT,
+        maxVolunteers: 0,
+      }, getBaseSetup('/api/shift/12'), [
+        {
+          path: ['maxVolunteers'],
+          message: 'Number must be greater than or equal to 1',
+        },
+      ]);
     });
-});
 
-test('PUT  /api/shift/-1; Shift fails because id is invalid', async () => {
-  await request(app)
-    .put('/api/shift/-1')
-    .send(myShift)
-    .expect(400)
-    .then((response) => {
-      expect(response.text).toBe('Invalid ID');
+    test('Shift fails because id is invalid', async () => {
+      await runValidationTestError(BASE_SHIFT, getBaseSetup('/api/shift/-1'), [
+        {
+          path: ['id'],
+          message: 'Number must be greater than or equal to 0',
+        },
+      ]);
     });
+  });
 });
```

Comment:
What about a scenario with multiple validation issues? Or does it error out at the first sign of trouble?

Reply:
My opinion here is that the actual implementation for Zod here doesn't super matter. I just am testing that the errors occur on the proper fields.

I believe that Zod does have the ability to abort early but it's sorta a hassle to get it setup to do so. I don't think testing with multiple validation issues adds much here. Could be wrong tho, just my opinion.